### PR TITLE
Multi-producer-multi-consumer queues

### PR DESCRIPTION
**Pull request by ThinkOpenly**
I'm not sure I've done all the git stuff correctly.  Apologies if I'm making a mess of things.  The important files for MPMC Qs are:
- configure
- src/Makefile.am
- src/Makefile.in
- src/sastype.h
- src/sphmultipcqueue.h
- src/sphmultipcqueue.cpp
- src/tests/sphmultipcqueue_t.c
- src/tests/sphmultipcqueue_ttt.c

### CODE REVIEW

**Comment by tuliom** [comment]
@ThinkOpenly , a suggestion for the next pull request: avoid adding merge commits to your pull request, because excessive merge commits makes the commit log more complex.  ;-)

File: src/sphmultipcqueue.cpp [hunk 1]
```diff
@@ -10,7 +10,7 @@
  *     IBM Corporation, Paul Clarke   - MPMC implementation
  */
 
-#define __SASDebugPrint__ 1
+//#define __SASDebugPrint__ 1
```
Comment (tuliom) [inline src/sphmultipcqueue.cpp]: Should this line be removed?

File: src/sphmultipcqueue.cpp [hunk 2]
```diff
@@ -421,24 +421,33 @@ SPHMultiPCQueueGetNextCompleteDirectTM (SPHMultiPCQueue_t queue)
...
 	}
+#if 0
+	{
+		SPHLFEntryHeader_t **tail_p = (SPHLFEntryHeader_t **)&headerBlock->qtail;
+		SPHLFEntryHeader_t *entryPtr = *tail_p;
+		sphLFEntry_t entrytemp;
+		entrytemp.idUnit = entryPtr->entryID.idUnit;
+		printf("%-6d: tail = %lx entryPtr = %p allocated = %d valid = %d\n",sphdeGetTID(),headerBlock->qtail,entryPtr,entrytemp.detail.allocated,entrytemp.detail.valid);
+	}
+#endif
```
Comment (tuliom) [inline src/sphmultipcqueue.cpp]: May we remove this block of code too?

File: src/sphmultipcqueue.h [hunk 3]
```diff
@@ -67,11 +67,11 @@ if (handle)
...
 *	The consumer:
-*	- checks (spins) for the next allocated entry to become complete.
-*	- uses the returned entry handle to directly access the entry contents.
+*	- checks (spins) for the next allocated entry to become complete
+*	- uses the returned entry handle to directly access the entry contents
 *	- When done processing the queue entry, it marks the entry header
-*	invalid and deallocates the entry.
-*	- This makes the next queue entry available, if any.
+*	invalid and deallocates the entry
+*	- This makes the next queue entry available, if any
```
Comment (tuliom) [inline src/sphmultipcqueue.h]: Are these changes really intentional?
Looks like you're removing valid periods '.' here.

File: src/Makefile.am [hunk 4]
```diff
@@ -1,7 +1,8 @@
 ## Process this file with autoconf to produce configure.
 
-AM_CFLAGS   = -Wall -Wwrite-strings $(PTHREAD_CFLAGS)
-AM_CXXFLAGS = -Wall -Wwrite-strings -Woverloaded-virtual $(PTHREAD_CFLAGS)
+AM_CFLAGS   = $(HTM) -fgnu-tm -Wall -Wwrite-strings $(PTHREAD_CFLAGS)
+AM_CXXFLAGS = $(HTM) -fgnu-tm -Wall -Wwrite-strings -Woverloaded-virtual $(PTHREAD_CFLAGS)
+LIBS = -litm
```
Comment (tuliom) [inline src/Makefile.am]: When you set LIBS this way, don't you break inheritance from the autoconf value?

File: src/Makefile.in [hunk 5]
```diff
@@ -348,8 +348,12 @@ SOURCES = $(libsphde_la_SOURCES) $(libsphgettime_la_SOURCES) \
 	$(sphgtod_t_SOURCES) $(sphlflogger_t_SOURCES) \
 	$(sphlflogger_tt_SOURCES) $(sphlflogger_ttt_SOURCES) \
 	$(sphlockfreeheap_t_SOURCES) $(sphlogportal_t_SOURCES) \
+<<<<<<< HEAD
 	$(sphlogportal_tt_SOURCES) $(sphmultipcqueue_t_SOURCES) \
 	$(sphmultipcqueue_ttt_SOURCES) $(sphsinglepcqueue_t_SOURCES) \
+=======
+	$(sphlogportal_tt_SOURCES) $(sphsinglepcqueue_t_SOURCES) \
+>>>>>>> 1a01901f9e99fc5b680697b081fea31139764d4a
```
Comment (tuliom) [inline src/Makefile.in]: Ooops.  You committed some merge conflicts here.

**Comment by ThinkOpenly** [comment]
> avoid adding merge commits to your pull request, because excessive merge commits makes the commit log more complex. ;-)

Apologies. I need more git education.  :-/

File: src/sphmultipcqueue.cpp (hunk 1 above)
Comment (ThinkOpenly) [inline src/sphmultipcqueue.cpp]: This is marked as a comment in "an outdated diff"?

I'd prefer to keep this line in, as an easy way to enable debug output.

File: src/sphmultipcqueue.cpp (hunk 2 above)
Comment (ThinkOpenly) [inline src/sphmultipcqueue.cpp]: yes, this can go.  sorry.

File: src/Makefile.am (hunk 4 above)
Comment (ThinkOpenly) [inline src/Makefile.am]: There is no "AM_LIBS". What do you recommend? Oh, maybe it should be "+=" ?

File: src/sphmultipcqueue.h (hunk 3 above)
Comment (ThinkOpenly) [inline src/sphmultipcqueue.h]: Technically, those periods are not needed, since the sentences are part of a bullet list. However, my changes are certainly not terribly important here. I defer to your preference.

File: src/Makefile.in (hunk 5 above)
Comment (ThinkOpenly) [inline src/Makefile.in]: I think I just fixed that with a fresh merge commit.

File: src/Makefile.am (hunk 4 above)
Comment (tuliom) [inline src/Makefile.am]: Using ```+=``` would work.
But ideally, you make this change in -litm in the configure.ac.  That's the only place that knows if libitm is really available in the system.

Comment (ThinkOpenly) [inline src/Makefile.am]: I just committed a change for this, using AC_SEARCH_LIBS. Of particular note is that configure will now _fail_ if libitm is not installed.

**Comment by tuliom** [comment]
@ThinkOpenly , I created a refactored branch on my fork where I removed all the merge commits from this pull request.  Could you take a look at it, please? https://github.com/tuliom/sphde/tree/refactor-pr26

**Comment by ThinkOpenly** [comment]
@tuliom, there's a patch or two more on my mpmc branch which eventually need to be integrated. It is primarily to enable multi-producer-single-consumer queues, plus some clean up and a more extensive test case.  If you want to do it on this branch, that's great. Let me know how I can help.  Also, my changes to "configure" need to be moved to whatever would generate the equivalent from "configure.ac", but I haven't figured that out yet (enabling HTM only with "--enable-htm" flag). Unfortunately, it'll have to wait for my return.

**Comment by RajalakshmiS** [comment]
@ThinkOpenly Is there a clean branch with all the commits which I can use for reviewing?

**Comment by ThinkOpenly** [comment]
@RajalakshmiS, let me work on that.  Also, I added the "--enable-htm" option poorly, so I'll fix that, too.

**Comment by tuliom** [comment]
Pull request #27 superseded this one.
